### DISCUSSION TITLE
Looking for solution for mutate dependency and multiple promises in mutate

### DISCUSSION THREAD

**Discussion by TommySorensen** [TanStack/query#3402] [Q&A]
Hi all.

I am currently building a reusable basket hook with React Query for the server state.
I have search the discussions and issues for examples of running a GET request before i want to call a POST request with `useMutation`, but haven't found any good examples of it.

Let's say that i 2 endpoints, one for creating(GET) a basket and one for updating(POST) it. The goal is to wait with creating the basket before i'm actually are putting something in it. So when user clicks "add to basket" i could in the `useMutation` do a fetch call, await it and then run the POST with the basket id i just created. This will properly work fine for first time running the mutate. But let's say the user can click the "add to basket" button multiple times, that will cause a issue since when clicking second time, the basket is not created and the previous promise i still running. So the question is: Are there anything that can handle cases like this i React Query?

I tried something like this:
```js
let basketId = somethingFromLocalStorage(); // will be undefined if nothing found in storage
        
return useMutation(
    async ({ quantity, itemNo }) => {
        let id = basketId;

        if (!id) {
            const { basketId } = await createBasket();
            id = basketId;
        }

        return addBasketItem(id, itemNo, quantity);
    },
    {
        onMutate: (variables) => {
            // Do some smart optimistic update with itemNo and quantity
        },
        onSuccess: () => {
            // Use queryClient.setQueryData to store data in a reusable basket hook
        },
    }
);
```
**User journey:**
Click "add to basket" button twice and expect the basket to contain a basket with quantity of 2 with only 2 requests, one for create and one for updating.

**Comment by ryanmk54** [comment]
You could have a queue, where if the mutation isLoading, the item gets stored in the queue instead of calling mutate. In the onSuccess, you could check the queue and mutate the variables in the queue if there is are variables waiting to be mutated.

**Reply by TommySorensen** [reply]
Let's say we create a queue of requests as you mentioned, then how would you connect those request to the basket id that is not created yet? I somehow need to attach the response from the first connect to the next queue requests while they keep coming in. If we do that logic in the onSuccess i still need to change the url of the requests before running them since they does not contain the basket id yet

**Reply by ryanmk54** [reply]
It sounds like maybe you need two mutations, one to create the basket, and one to add the items. Here is a related discussion: https://github.com/TkDodo/blog-comments/issues/31#issuecomment-939810580

I would do something similar to the code below:
```
const [itemsToAdd, setItemsToAdd] = useState([])
const createBasketMutation = useMutation({
  onSuccess: (basket) => {
    itemsToAdd.forEach((item) => {
      // basket id was returned from server in the basket variable
      const url = `/items/${basket.id}`
    })
    setItemsToAdd([])
  }
})
const addItemMutation = useMutation({
  mutationFn: (item) => {
    if (createBasketMutation.data) {
      // basket id is in createBasketMutation.data.id
    }
    else {
      setItemsToAdd(prev => [...prev, item])
    }
  }
})
```

**Reply by TommySorensen** [reply]
Thx. for helping. I thing i got it working okay right now with use of local storage sync where i watch for storage changes. I did try with hooks, bit it didn't right get it working since i then got 2 caches of the basket and they need to update each other.

```js
function useBasket() {
    const [storedReference, setStoredReference] = useBasketReference();

    return useQuery(
        ['basket'],
        (context) => {
            const { meta } = context;
            const reference = meta?.reference;

            if (reference) {
                return getBasket(reference);
            }

            return createBasket();
        },
        {
            refetchOnWindowFocus: true,
            onSuccess: (data) => {
                if (!storedReference || storedReference !== data?.basketReference) {
                    setStoredReference(data.basketReference);
                }
            },
            meta: {
                reference: storedReference,
            },
        }
    );
}
```

I switched the code so i now have one hook called `useBasket` that can create the basket on browser load, and when it's created the same hook will now get the basket. So i can now other places use `useMutation` with `['basket']` key to update it instead of having 2 keys with tha basket data in. And the data is synced between tabs because of using a useLocalStorage hook helper. So im stuck with creating a new basket for every user on the site to avoid having the queue thing.

**Reply by TommySorensen** [reply]
@ryanmk54 Okay, so this didn't really solve it since both onSuccess and onError will fire for every time i use the useBasket hook 🤔

**Reply by ryanmk54** [reply]
What's the error message and stack trace?

**Reply by TommySorensen** [reply]
What I mean is that since the useBasket hook is a shared hook I use a lot of places. Then the onSuccess will be called on every hook whenever the fetch is returning success. So maybe I should look into the global subscribe event instead on that query key

**Reply by TkDodo** [reply]
if you want callbacks (onSuccess / onError / onSettled) that are only called once per query, you are looking for the [global cache callbacks](https://react-query.tanstack.com/reference/QueryCache#global-callbacks). I've also written about them (to some extent) [here](https://tkdodo.eu/blog/react-query-error-handling#the-onerror-callback).

**Reply by TommySorensen** [reply]
@TkDodo Thx. That is also the solution im looking into right now by using the `QueryObserver` with subscribe or `QueryCache`. But i still don't have a good solution for my initial question since it might require a queue of fetch calls that should await the first request to finish.
